rioxarray/raster_index.py

```diff
@@ -0,0 +1,428 @@
+            }
+        else:
+            xy_transform = AffineTransform(
+                affine, width, height, x_dim=x_dim, y_dim=y_dim
+            )
+            indexes = {("x", "y"): CoordinateTransformIndex(xy_transform)}
+
+        return cls(indexes)
+
+    @classmethod
+    def from_variables(
+        cls,
+        variables: Mapping[Any, Variable],
+        *,
+        options: Mapping[str, Any],
+    ) -> "RasterIndex":
+        # TODO: compute bounds, resolution and affine transform from explicit coordinates.
+        raise NotImplementedError(
+            "Creating a RasterIndex from existing coordinates is not yet supported."
+        )
+
+    def create_variables(
+        self, variables: Mapping[Any, Variable] | None = None
+    ) -> dict[Hashable, Variable]:
+        new_variables: dict[Hashable, Variable] = {}
+
+        for index in self._wrapped_indexes.values():
+            new_variables.update(index.create_variables())
+
+        return new_variables
+
+    def isel(
+        self, indexers: Mapping[Any, int | slice | np.ndarray | Variable]
+    ) -> "RasterIndex | None":
+        new_indexes: dict[WrappedIndexCoords, WrappedIndex] = {}
+
+        for coord_names, index in self._wrapped_indexes.items():
+            index_indexers = _filter_dim_indexers(index, indexers)
+            if not index_indexers:
+                # no selection to perform: simply propagate the index
+                # TODO: uncomment when https://github.com/pydata/xarray/issues/10063 is fixed
+                # new_indexes[coord_names] = index
+                ...
+            else:
+                new_index = index.isel(index_indexers)
+                if new_index is not None:
+                    new_indexes[coord_names] = new_index
+
+        if new_indexes:
+            # TODO: if there's only a single PandasIndex can we just return it?
+            # (maybe better to keep it wrapped if we plan to later make RasterIndex CRS-aware)
+            return RasterIndex(new_indexes)
+        else:
+            return None
+
+    def sel(
+        self, labels: dict[Any, Any], method=None, tolerance=None
+    ) -> IndexSelResult:
+        results = []
+
+        for coord_names, index in self._wrapped_indexes.items():
+            if not isinstance(coord_names, tuple):
+                coord_names = (coord_names,)
+            index_labels = {k: v for k, v in labels if k in coord_names}
+            if index_labels:
+                results.append(
+                    index.sel(index_labels, method=method, tolerance=tolerance)
+                )
+
+        return merge_sel_results(results)
+
+    def equals(self, other: Index) -> bool:
+        if not isinstance(other, RasterIndex):
+            return False
+        if set(self._wrapped_indexes) != set(other._wrapped_indexes):
+            return False
+
+        return all(
+            index.equals(other._wrapped_indexes[k])  # type: ignore[arg-type]
```

Comment:
Need to relax the typing constraints of `xarray.core.indexes.CoordinateTransformIndex.equals()` (other should be any `xarray.Index`) so we can remove the `# type: ignore` here.